File: atsa/src/main/java/jp/kaiz/atsassistmod/controller/trainprotection/RATSController.java
package jp.kaiz.atsassistmod.controller.trainprotection;

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity;

/**
 * R-ATS uses the signal aspect ahead of the train. RTM's old {@code getSignal()}
 * is a stub in RealTrainModRenewed (returns 0), so until the new signal network
 * is bridged this resolves to "no limit". Behaviour is otherwise faithful.
 */
public class RATSController extends TrainProtection {
    private int limitSpeed;

    @Override
    public void onTick(TrainEntity train, double distance) throws Exception {
        super.onTick(train, distance);
        switch (train.getSignal()) {
            case 2 -> this.limitSpeed = 30;
            case 3 -> this.limitSpeed = 45;
            case 4 -> this.limitSpeed = 65;
            case 5 -> this.limitSpeed = 95;
            default -> this.limitSpeed = Integer.MAX_VALUE;
        }
    }

    @Override
    public int getNotch(float speedH) {
        if (this.limitSpeed == Integer.MAX_VALUE) {
            return 1;
        }
        float overSpeed = speedH - this.limitSpeed;
        if (overSpeed > 5) {
            return -7;
        } else if (overSpeed > 0) {
            return -4;
        } else {
            return 1;
        }
    }

    @Override
    public TrainProtectionType getType() {
        return TrainProtectionType.RATS;
    }

    @Override
    public int getDisplaySpeed() {
        return this.limitSpeed;
    }
}

File: atsa/src/main/kotlin/jp/kaiz/atsassistmod/controller/trainprotection/RATSController.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package jp.kaiz.atsassistmod.controller.trainprotection

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity

/** R-ATS uses the signal aspect ahead of the train through RTM's legacy getSignal compatibility API. */
open class RATSController : TrainProtection() {
    private var limitSpeed = 0

    @Throws(Exception::class)
    override fun onTick(train: TrainEntity, distance: Double) {
        super.onTick(train, distance)
        limitSpeed = when (train.signal) {
            2 -> 30
            3 -> 45
            4 -> 65
            5 -> 95
            else -> Int.MAX_VALUE
        }
    }

    override fun getNotch(speedH: Float): Int {
        if (limitSpeed == Int.MAX_VALUE) {
            return 1
        }
        val overSpeed = speedH - limitSpeed
        return when {
            overSpeed > 5 -> -7
            overSpeed > 0 -> -4
            else -> 1
        }
    }

    override fun getType(): TrainProtectionType = TrainProtectionType.RATS

    override fun getDisplaySpeed(): Int = limitSpeed
}
